# Лабораторная работа 3.1

## Счастливый билетик

![](http://f2.mylove.ru/UNiHT7dRJ4.jpg)

Билет считается счастливым, если выполнено следующее условие: сумма первых трёх цифр номера равна сумме последних трёх цифр.

**Задание:**

1) Определите число `ticket_number` — шестизначный номер билета;

2) Напишите код, который по шестизначному номеру `ticket_number` билетика проверяет, является ли он счастливым;

3) Если номер счастливый, выведите строку `Yes`, иначе — `No`.

---

**Пример 1:**

*Input:* 123456

*Output:* No


**Пример 2:**

*Input:* 123042

*Output:* Yes

In [1]:
ticket_number = int(input("Enter the number on your ticket:"))
ahead_3 = ticket_number // 1000
behind_3 = ticket_number % 1000
if (ahead_3 // 100 + ahead_3 // 10 % 10 + ahead_3 % 10 == behind_3 // 100 + behind_3 // 10 % 10 + behind_3 % 10):
  print("Yes")
else:
  print("No")

Enter the number on your ticket:123321
Yes


## Пароль

![](http://forum.ixbt.com/post.cgi?id=attach:77:14138:1341:1)

Пусть пароль может содержать только латинские буквы, знаки препинания и цифры.

Пароль считается надёжным, если удовлетворяет следующим условиям:
- содержит буквы в разных регистрах;
- содержит цифры;
- содержит не менее `4` уникальных символов;
- не содержит ваше имя латинницей, записанное буквами любых регистров (anna, iVan, ...).

Иначе пароль считается слабым.

**Задание:**

1) Определите строку `password` — придуманный вами пароль;

2) Напишите код, который по паролю `password` проверяет, является ли он надёжным;

3) Если пароль надёжный, выведите строку `strong`, иначе — `weak`.


---

Пусть имя пользоватея -- Андрей.

**Пример 1:**

*Input:* Aandrei123

*Output:* weak


**Пример 2:**

*Input:* an12dRei

*Output:* strong

In [1]:
password = input("Enter your password: ")
name = input("Enter your name: ")
if (password == password.upper() or password == password.lower() or password.isalpha() 
   or len(set(password)) < 4 or name.lower() in password.lower()):
  print("your password is weak")
else:
  print("your password is strong")

Enter your password:  Aandrei123
Enter your name: Andrey
your password is strong


## Числа Фибоначчи
Как известно, [числа Фибоначчи](https://ru.wikipedia.org/wiki/%D0%A7%D0%B8%D1%81%D0%BB%D0%B0_%D0%A4%D0%B8%D0%B1%D0%BE%D0%BD%D0%B0%D1%87%D1%87%D0%B8) — это последовательность чисел, каждое из которых равно сумме двух предыдущих (первые два числа равны $1$):
$$
    1, 1, 2, 3, 5, 8, 13, \ldots
$$

**Задание:**

1) Определите число `amount` — количество чисел Фибоначчи, которые надо вывести;

2) Напишите код, который выводит первые `amount` чисел Фибоначчи.

---

**Пример 1:**

*Input:* 3

*Output:* 1 1 2


**Пример 2:**

*Input:* 10

*Output:* 1 1 2 3 5 8 13 21 34 55


In [3]:
amount = int(input("Enter a digit: "))
a, b = 0, 1
print(b)
for i in range(amount):
    sum = a + b
    a = b
    b = sum
    print(b)

Enter a digit: 3
1
1
2
3


## Время исследований

![](https://2.bp.blogspot.com/-VOT5x12UqPY/WSh53tQ94HI/AAAAAAAAAMY/-nlcz28_Fj4UzJjbn4bBINKCnzwag1W4QCLcB/s1600/my-pic-widget.jpg)

На сайте https://www.kaggle.com/ выберите любой набор данных в формате CSV и проведите для него маленькое исследование: загрузите данные из набора с использованием стандартного модуля `csv`, посмотрите средние значения и стандартные отклонения двух выбранных числовых атрибутов, найдите [методом наименьших квадратов](https://ru.wikipedia.org/wiki/%D0%9C%D0%B5%D1%82%D0%BE%D0%B4_%D0%BD%D0%B0%D0%B8%D0%BC%D0%B5%D0%BD%D1%8C%D1%88%D0%B8%D1%85_%D0%BA%D0%B2%D0%B0%D0%B4%D1%80%D0%B0%D1%82%D0%BE%D0%B2) уравнение линейной зависимости, связывающей один числовой атрибут с другим. Для оценки заданной зависимости найдите [коэффициент парной корреляции](https://ru.wikipedia.org/wiki/%D0%9A%D0%BE%D1%80%D1%80%D0%B5%D0%BB%D1%8F%D1%86%D0%B8%D1%8F), сделайте соответствующие выводы.

**Результаты надо обязательно прокомментировать и пояснить!**

---

**Пример 1:**

Пусть таблица `bikes.csv` содержит данные по арендам велосипедов за $2$ года:
- `datetime`: дата и время аренды
- `season`: время года
- `temp`: температура воздуха по Цельсию
- `windspeed`: скорость ветра 
- `registered`: число аренд

Одно из направлений исследования могло бы заключаться в проверке зависимости суммарного числа аренд от температуры воздуха.

Импорт библиотек:

Чтение файла с данными и вывод столбцов:

In [22]:
import csv
from math import sqrt


with open('avocado.csv', 'r', newline='') as csvf:
    data = csv.reader(csvf, delimiter=',')
    sml_b = []
    lrg_b = []
    for row in data:
        if row[8] == "Small Bags":
            continue
        sml_b.append(float(row[8]))
        lrg_b.append(float(row[9]))

*calculation the average value in 2 columns*

In [30]:
s_b = sum(sml_b) / len(sml_b)
l_b = sum(lrg_b) / len(lrg_b)
print(f"average meaning for Small Bags {s_b}")
print(f"average meaning for Large Bags : {l_b}")


average meaning for Small Bags 182194.68669571026
average meaning for Large Bags : 54338.08814455636


*calculation the standard deviations of the two selected numeric attributes*

In [27]:
s1 = sum((component-s_b)**2 for component in sml_b) / len(sml_b)
sd_s = sqrt(s1)
l2 = sum((component1-l_b)**2 for component1 in lrg_b) / len(lrg_b)
sd_l = sqrt(l2)
print(f"standard deviation for Small Bags: {sd_s}")
print(f"standard deviation for Large_Bags: {sd_l}")

standard deviation for Small Bags: 746158.0703144785
standard deviation for Large_Bags: 243959.28009055112


*The equation of linear dependence*

In [28]:
sum_ab = 0
sum_square = 0

for i, component in enumerate(sml_b):
    sum_ab += component * sml_b[i]
    sum_square += component ** 2

size = len(sml_b)
k_line = (size * sum_ab - sum(sml_b) * sum(lrg_b))/(size * sum_square - sum(sml_b)**2)
b_line = s_b - l_b * k_lin
func = []

for component in sml_b:
    func.append(k_line * component + b_line)

print(f"The equation of linear dependence: y = {k_line}x + {b_line}")

The equation of linear dependence: y = 1.0418405295823145x + 155170.31282014563


*The pair correlation coefficient:*

In [29]:
cor_dig = 0
for i, component in enumerate(sml_b):
    cor_dig += (component - s_b) * (lrg_b[i] - l_b)
sqr_diff_s = sum((comp - s_b)**2 for comp in sml_b)
sqr_diff_l = sum((comp - l_b)**2 for comp in lrg_b)
r_xy = cor_dig / sqrt(sqr_diff_s * sqr_diff_l)
print(f"Pair correlation coefficient: {r_xy}")

Pair correlation coefficient: 0.9025888652939013


**The pair correlation coefficient is ~ 0.90, which may indicate an implicit dependence**